test/test-lint.cpp

```diff
@@ -3005,6 +3013,146 @@ TEST(test_lint_eval, disable_variable_lookup_in_presence_of_eval) {
+    l.visit_variable_use(identifier_of(use_eval));
+    l.visit_variable_declaration(identifier_of(parameter_declaration),
+                                 variable_kind::_parameter);
+    l.visit_enter_function_scope_body();
+    l.visit_variable_use(identifier_of(use));
+    l.visit_exit_function_scope();
+    l.visit_variable_use(identifier_of(use));
+    l.visit_end_of_module();
+
+    EXPECT_THAT(v.errors,
+                ElementsAre(ERROR_TYPE_FIELD(error_use_of_undeclared_variable,
+                                             name, span_matcher(use))));
+  }
+
+  {
+    const char8 eval_declaration[] = u8"eval";
+
+    // let eval = console.log;
+    // eval("var x = 42;");
+    // x;  // ERROR (use of undeclared variable)
+    // x = 10;  // ERROR (assignment to undeclared variable)
+    error_collector v;
+    linter l(&v);
+    l.visit_variable_declaration(identifier_of(eval_declaration),
+                                 variable_kind::_let);
+    l.visit_variable_use(identifier_of(use_eval));
+    l.visit_variable_use(identifier_of(use));
+    l.visit_variable_assignment(identifier_of(use));
+    l.visit_end_of_module();
+
+    EXPECT_THAT(
+        v.errors,
+        ElementsAre(ERROR_TYPE_FIELD(error_use_of_undeclared_variable, name,
+                                     span_matcher(use)),
+                    ERROR_TYPE_FIELD(error_assignment_to_undeclared_variable,
+                                     assignment, span_matcher(use))));
+  }
+
+  {
+    const char8 eval_declaration[] = u8"eval";
+
+    // function f() {
+    //   eval = console.log;
+    //   eval("var x = 42;");
+    //   x;  // ERROR (use of undeclared variable)
+    //   x = 10;  // ERROR (assignment to undeclared variable)
+    //   {
+    //     var eval;
+    //   }
+    // }
+    error_collector v;
+    linter l(&v);
+    l.visit_enter_function_scope();
+    l.visit_enter_function_scope_body();
+    l.visit_variable_assignment(identifier_of(use_eval));
+    l.visit_variable_use(identifier_of(use_eval));
+    l.visit_variable_use(identifier_of(use));
+    l.visit_variable_assignment(identifier_of(use));
+    l.visit_enter_block_scope();
+    l.visit_variable_declaration(identifier_of(eval_declaration),
+                                 variable_kind::_var);
+    l.visit_exit_block_scope();
+    l.visit_exit_function_scope();
+    l.visit_end_of_module();
+
+    EXPECT_THAT(
+        v.errors,
+        ElementsAre(ERROR_TYPE_FIELD(error_use_of_undeclared_variable, name,
+                                     span_matcher(use)),
+                    ERROR_TYPE_FIELD(error_assignment_to_undeclared_variable,
+                                     assignment, span_matcher(use))));
+  }
+}
+
+TEST(test_lint_eval, false_negatives_on_redeclaration_of_eval) {
+  const char8 use_eval[] = u8"eval";
+  const char8 use[] = u8"x";
+
+  {
+    const char8 eval_declaration[] = u8"eval";
+
+    // let eval = console.log;
```

Comment:
Nit: Visits don't match the code.
```suggestion
    // let eval = () => {};
```